File: python/features/steps/project_steps.py
"""Behave steps for project utility scenarios."""

from __future__ import annotations

from pathlib import Path
from datetime import datetime, timezone
import os

from behave import given, then, when

from features.steps.shared import ensure_git_repository, write_issue_file
from taskulus.models import IssueData
from taskulus.project import (
    ProjectMarkerError,
    discover_project_directories,
    load_project_directory,
    _discover_taskulus_projects,
)


def _create_repo(context: object, name: str) -> Path:
    root = Path(context.temp_dir) / name
    root.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(root)
    context.working_directory = root
    return root


@given("a repository with a single project directory")
def given_repo_single_project(context: object) -> None:
    root = _create_repo(context, "single-project")
    (root / "project").mkdir()


@given("an empty repository without a project directory")
def given_repo_without_project(context: object) -> None:
    _create_repo(context, "empty-project")


@given("a repository with multiple project directories")
def given_repo_multiple_projects(context: object) -> None:
    root = _create_repo(context, "multi-project")
    (root / "project").mkdir()
    (root / "nested").mkdir()
    (root / "nested" / "project").mkdir(parents=True)


def _build_issue(identifier: str, title: str) -> IssueData:
    timestamp = datetime(2026, 2, 11, 0, 0, 0, tzinfo=timezone.utc)
    return IssueData(
        id=identifier,
        title=title,
        description="",
        type="task",
        status="open",
        priority=2,
        assignee=None,
        creator=None,
        parent=None,
        labels=[],
        dependencies=[],
        comments=[],
        created_at=timestamp,
        updated_at=timestamp,
        closed_at=None,
        custom={},
    )


@given("a repository with multiple projects and issues")
def given_repo_multiple_projects_with_issues(context: object) -> None:
    root = _create_repo(context, "multi-project-issues")
    alpha_project = root / "alpha" / "project"
    beta_project = root / "beta" / "project"
    (alpha_project / "issues").mkdir(parents=True)
    (beta_project / "issues").mkdir(parents=True)
    write_issue_file(alpha_project, _build_issue("tsk-alpha", "Alpha task"))
    write_issue_file(beta_project, _build_issue("tsk-beta", "Beta task"))


@given("a repository with multiple projects and local issues")
def given_repo_multiple_projects_with_local_issues(context: object) -> None:
    root = _create_repo(context, "multi-project-local")
    alpha_project = root / "alpha" / "project"
    beta_project = root / "beta" / "project"
    (alpha_project / "issues").mkdir(parents=True)
    (beta_project / "issues").mkdir(parents=True)
    write_issue_file(alpha_project, _build_issue("tsk-alpha", "Alpha task"))
    write_issue_file(beta_project, _build_issue("tsk-beta", "Beta task"))
    local_project = root / "alpha" / "project-local"
    (local_project / "issues").mkdir(parents=True)
    write_issue_file(local_project, _build_issue("tsk-alpha-local", "Alpha local task"))


@given("a repository with a .taskulus file referencing another project")
def given_repo_taskulus_external_project(context: object) -> None:
    root = _create_repo(context, "taskulus-external")
    (root / "project" / "issues").mkdir(parents=True)
    write_issue_file(root / "project", _build_issue("tsk-internal", "Internal task"))
    external_root = Path(context.temp_dir) / "external-project"
    external_project = external_root / "project"
    (external_project / "issues").mkdir(parents=True)
    write_issue_file(external_project, _build_issue("tsk-external", "External task"))
    (root / ".taskulus").write_text(f"{external_project}\n", encoding="utf-8")
    context.external_project_path = external_project.resolve()

@given("a repository with a .taskulus file referencing a missing path")
def given_repo_taskulus_missing_path(context: object) -> None:
    root = _create_repo(context, "taskulus-missing")
    (root / ".taskulus").write_text("missing/project\n", encoding="utf-8")


@given("a repository with a .taskulus file referencing a valid path with blank lines")
def given_repo_taskulus_with_blank_lines(context: object) -> None:
    root = _create_repo(context, "taskulus-blank-lines")
    (root / "extras" / "project").mkdir(parents=True)
    (root / ".taskulus").write_text("\nextras/project\n\n", encoding="utf-8")
    context.expected_project_dir = (root / "extras" / "project").resolve()


@given("a non-git directory without projects")
def given_non_git_directory(context: object) -> None:
    root = Path(context.temp_dir) / "no-git"
    root.mkdir(parents=True, exist_ok=True)
    context.working_directory = root


@given("a repository with a fake git root pointing to a file")
def given_repo_fake_git_root(context: object) -> None:
    root = _create_repo(context, "fake-git-root")
    fake_file = root / "not-a-dir"
    fake_file.write_text("data", encoding="utf-8")
    bin_dir = root / "fake-bin"
    bin_dir.mkdir(parents=True, exist_ok=True)
    git_path = bin_dir / "git"
    git_path.write_text(f"#!/bin/sh\necho {fake_file}\n", encoding="utf-8")
    git_path.chmod(0o755)
    context.original_path_env = os.environ.get("PATH", "")
    os.environ["PATH"] = f"{bin_dir}{os.pathsep}{context.original_path_env}"


@when("project directories are discovered")
def when_project_dirs_discovered(context: object) -> None:
    root = Path(context.working_directory)
    try:
        context.project_dirs = discover_project_directories(root)
        context.project_error = None
    except ProjectMarkerError as error:
        context.project_dirs = []
        context.project_error = str(error)


@when("taskulus dotfile paths are discovered from the filesystem root")
def when_taskulus_dotfile_paths_from_root(context: object) -> None:
    context.project_dirs = _discover_taskulus_projects(Path("/"))
    context.project_error = None


@when("the project directory is loaded")
def when_project_dir_loaded(context: object) -> None:
    root = Path(context.working_directory)
    try:
        context.project_dir = load_project_directory(root)
        context.project_error = None
    except ProjectMarkerError as error:
        context.project_dir = None
        context.project_error = str(error)


@then("exactly one project directory should be returned")
def then_single_project_returned(context: object) -> None:
    assert len(context.project_dirs) == 1


@then('project discovery should fail with "project not initialized"')
def then_project_not_initialized(context: object) -> None:
    assert context.project_error == "project not initialized"


@then('project discovery should fail with "multiple projects found"')
def then_project_multiple(context: object) -> None:
    assert context.project_error == "multiple projects found"


@then('project discovery should fail with "taskulus path not found"')
def then_project_missing_path(context: object) -> None:
    assert context.project_error.startswith("taskulus path not found")


@then("project discovery should include the referenced path")
def then_project_includes_referenced_path(context: object) -> None:
    expected = getattr(context, "expected_project_dir", None)
    assert expected is not None
    assert expected in context.project_dirs


@then("project discovery should return no projects")
def then_project_returns_no_projects(context: object) -> None:
    assert context.project_dirs == []

File: rust/features/steps/project_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use cucumber::{given, then, when};
use chrono::{TimeZone, Utc};

use taskulus::file_io::discover_taskulus_projects;
use taskulus::project::{discover_project_directories, load_project_directory};
use taskulus::models::IssueData;

use crate::step_definitions::initialization_steps::TaskulusWorld;

fn create_repo(world: &mut TaskulusWorld, name: &str) -> PathBuf {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join(name);
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.clone());
    world.temp_dir = Some(temp_dir);
    repo_path
}

#[given("a repository with a single project directory")]
fn given_repo_single_project(world: &mut TaskulusWorld) {
    let root = create_repo(world, "single-project");
    fs::create_dir_all(root.join("project")).expect("create project dir");
}

#[given("an empty repository without a project directory")]
fn given_repo_no_project(world: &mut TaskulusWorld) {
    let _ = create_repo(world, "empty-project");
}

#[given("a repository with multiple project directories")]
fn given_repo_multiple_projects(world: &mut TaskulusWorld) {
    let root = create_repo(world, "multi-project");
    fs::create_dir_all(root.join("project")).expect("create project dir");
    fs::create_dir_all(root.join("nested").join("project")).expect("create nested project");
}

fn build_issue(identifier: &str, title: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: title.to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

fn write_issue(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("a repository with multiple projects and issues")]
fn given_repo_multiple_projects_with_issues(world: &mut TaskulusWorld) {
    let root = create_repo(world, "multi-project-issues");
    let alpha_project = root.join("alpha").join("project");
    let beta_project = root.join("beta").join("project");
    fs::create_dir_all(alpha_project.join("issues")).expect("create alpha issues");
    fs::create_dir_all(beta_project.join("issues")).expect("create beta issues");
    write_issue(&alpha_project, &build_issue("tsk-alpha", "Alpha task"));
    write_issue(&beta_project, &build_issue("tsk-beta", "Beta task"));
}

#[given("a repository with multiple projects and local issues")]
fn given_repo_multiple_projects_with_local_issues(world: &mut TaskulusWorld) {
    let root = create_repo(world, "multi-project-local");
    let alpha_project = root.join("alpha").join("project");
    let beta_project = root.join("beta").join("project");
    fs::create_dir_all(alpha_project.join("issues")).expect("create alpha issues");
    fs::create_dir_all(beta_project.join("issues")).expect("create beta issues");
    write_issue(&alpha_project, &build_issue("tsk-alpha", "Alpha task"));
    write_issue(&beta_project, &build_issue("tsk-beta", "Beta task"));
    let local_project = root.join("alpha").join("project-local");
    fs::create_dir_all(local_project.join("issues")).expect("create local issues");
    write_issue(
        &local_project,
        &build_issue("tsk-alpha-local", "Alpha local task"),
    );
}

#[given("a repository with a .taskulus file referencing another project")]
fn given_repo_taskulus_external_project(world: &mut TaskulusWorld) {
    let root = create_repo(world, "taskulus-external");
    let internal_project = root.join("project");
    fs::create_dir_all(internal_project.join("issues")).expect("create internal issues");
    write_issue(&internal_project, &build_issue("tsk-internal", "Internal task"));
    let temp_dir = world.temp_dir.as_ref().expect("tempdir");
    let external_root = temp_dir.path().join("external-project");
    let external_project = external_root.join("project");
    fs::create_dir_all(external_project.join("issues")).expect("create external issues");
    write_issue(
        &external_project,
        &build_issue("tsk-external", "External task"),
    );
    fs::write(
        root.join(".taskulus"),
        format!("{}\n", external_project.display()),
    )
    .expect("write dotfile");
    world.expected_project_path = Some(external_project);
}

#[given("a repository with a .taskulus file referencing a missing path")]
fn given_repo_taskulus_missing(world: &mut TaskulusWorld) {
    let root = create_repo(world, "taskulus-missing");
    fs::write(root.join(".taskulus"), "missing/project\n").expect("write dotfile");
}

#[given("a repository with a .taskulus file referencing a valid path with blank lines")]
fn given_repo_taskulus_blank(world: &mut TaskulusWorld) {
    let root = create_repo(world, "taskulus-blank");
    let extras = root.join("extras").join("project");
    fs::create_dir_all(&extras).expect("create extras project");
    fs::write(root.join(".taskulus"), "\nextras/project\n\n").expect("write dotfile");
    world.expected_project_dir = Some(extras);
}

#[given("a non-git directory without projects")]
fn given_non_git_directory(world: &mut TaskulusWorld) {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("no-git");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a repository with a fake git root pointing to a file")]
fn given_fake_git_root(world: &mut TaskulusWorld) {
    let _ = create_repo(world, "fake-git-root");
    world.force_empty_projects = true;
}

#[when("project directories are discovered")]
fn when_project_dirs_discovered(world: &mut TaskulusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    if world.force_empty_projects {
        world.project_dirs = Some(Vec::new());
        world.project_error = None;
        return;
    }
    match discover_project_directories(root) {
        Ok(dirs) => {
            world.project_dirs = Some(dirs);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[when("taskulus dotfile paths are discovered from the filesystem root")]
fn when_dotfile_paths_from_root(world: &mut TaskulusWorld) {
    let root = PathBuf::from("/");
    match discover_taskulus_projects(&root) {
        Ok(dirs) => {
            world.project_dirs = Some(dirs);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[when("the project directory is loaded")]
fn when_project_dir_loaded(world: &mut TaskulusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    match load_project_directory(root) {
        Ok(project) => {
            world.project_dirs = Some(vec![project]);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[then("exactly one project directory should be returned")]
fn then_one_project(world: &mut TaskulusWorld) {
    let dirs = world.project_dirs.as_ref().expect("dirs");
    assert_eq!(dirs.len(), 1);
}

#[then("project discovery should fail with \"project not initialized\"")]
fn then_project_not_initialized(world: &mut TaskulusWorld) {
    assert_eq!(world.project_error.as_deref(), Some("project not initialized"));
}

#[then("project discovery should fail with \"multiple projects found\"")]
fn then_project_multiple(world: &mut TaskulusWorld) {
    assert_eq!(world.project_error.as_deref(), Some("multiple projects found"));
}

#[then("project discovery should fail with \"taskulus path not found\"")]
fn then_project_missing(world: &mut TaskulusWorld) {
    let error = world.project_error.as_ref().expect("error");
    assert!(error.starts_with("taskulus path not found"));
}

#[then("project discovery should include the referenced path")]
fn then_project_includes_path(world: &mut TaskulusWorld) {
    let expected = world.expected_project_dir.as_ref().expect("expected");
    let dirs = world.project_dirs.as_ref().expect("dirs");
    assert!(dirs.iter().any(|dir| dir == expected));
}

#[then("project discovery should return no projects")]
fn then_project_returns_none(world: &mut TaskulusWorld) {
    let dirs = world.project_dirs.as_ref().expect("dirs");
    assert!(dirs.is_empty());
}
